lib/cache/memory-cache-store.js

```diff
@@ -0,0 +1,417 @@
+      //  property.
+
+      if (this.isFull) {
+        // Or not, we don't have space to add another response
+        return undefined
+      }
+
+      this.#entryCount++
+
+      value = {
+        readers: 0,
+        readLock: false,
+        writeLock: false,
+        opts,
+        body: []
+      }
+
+      // We want to sort our responses in decending order by their deleteAt
+      //  timestamps so that deleting expired responses is faster
+      if (
+        values.length === 0 ||
+        opts.deleteAt < values[values.length - 1].deleteAt
+      ) {
+        // Our value is either the only response for this path or our deleteAt
+        //  time is sooner than all the other responses
+        values.push(value)
+      } else if (opts.deleteAt >= values[0].deleteAt) {
+        // Our deleteAt is later than everyone elses
+        values.unshift(value)
+      } else {
+        // We're neither in the front or the end, let's just binary search to
+        //  find our stop we need to be in
+        let startIndex = 0
+        let endIndex = values.length
+        while (true) {
+          if (startIndex === endIndex) {
+            values.splice(startIndex, 0, value)
+            break
+          }
+
+          const middleIndex = Math.floor((startIndex + endIndex) / 2)
+          const middleValue = values[middleIndex]
+          if (opts.deleteAt === middleIndex) {
+            values.splice(middleIndex, 0, value)
+            break
+          } else if (opts.deleteAt > middleValue.opts.deleteAt) {
+            endIndex = middleIndex
+            continue
+          } else {
+            startIndex = middleIndex
+            continue
+          }
+        }
+      }
+    } else {
+      // Check if there's already another request writing to the value or
+      //  a request reading from it
+      if (value.writeLock || value.readLock) {
+        return undefined
+      }
+
+      // Empty it so we can overwrite it
+      value.body = []
+    }
+
+    const writable = new MemoryStoreWritableStream(
+      value,
+      this.#maxEntrySize
+    )
+
+    // Remove the value if there was some error
+    writable.on('error', (err) => {
+      values.filter(current => value !== current)
+      if (this.#errorCallback) {
+        this.#errorCallback(err)
+      }
+    })
+
+    writable.on('bodyOversized', () => {
+      values.filter(current => value !== current)
+    })
+
+    return writable
+  }
+
+  /**
+   * @param {string} origin
+   */
+  deleteByOrigin (origin) {
+    this.#data.delete(origin)
+  }
+
+  /**
+   * Gets all of the requests of the same origin, path, and method. Does not
+   *  take the `vary` property into account.
+   * @param {import('../../types/dispatcher.d.ts').default.RequestOptions} req
+   * @returns {MemoryStoreValue[] | undefined}
+   */
+  #getValuesForRequest (req, makeIfDoesntExist) {
+    // https://www.rfc-editor.org/rfc/rfc9111.html#section-2-3
+    let cachedPaths = this.#data.get(req.origin)
```

Comment:
is req.origin always set?

Reply:
at this stage yes - we verify its existence in the cache interceptor